fuse/utils/interface_validator.py

```diff
@@ -0,0 +1,156 @@
+    def draw(self, *, color:str = 'red', style:int=12, **kwargs):
+        print(f'FruitBase.draw: {color=} {style=}')
+
+    @validate_signature(args=['size', 'material']) #both args `size` and `material` will be validated
+    def print_3d(self, *, size:float = 2.7, material:str = 'iron'):
+        print(f'FruitBase.print_3d {size=} {material=}')
+```
+
+now, when you use a class that inherits from it, like:
+
+```python
+class Banana(FruitBase):
+    #...
+```
+
+the API will be enforced and exceptions will be raised if requirements are not met.
+
+
+#TODO:
+    1. allow validating return type hint
+    2. allow using @validate_signature without using any args, and that would mean that the exact same full signature is required
+    3. allow control over whether type hint is validated or not (is this needed?)
+"""
+
+
+def validate_signature(*, args: List[str] = None) -> Callable:
+    def wrapper(function: Callable) -> Any:
+        return validate_signature_impl(function, args=args)
+
+    return wrapper
+
+
+class validate_signature_impl(object):
+    def __init__(self, fn: Callable, args: List[str] = None) -> None:
+        self.fn = fn
+        self._args_to_validate = args
+
+    def __set_name__(self, owner: any, name: str) -> None:
+        # do something with owner, i.e.
+        curr_func = self.fn
+        print(f"decorating {curr_func} and using {owner}")
+        if not hasattr(owner, "_validate_methods"):
+            owner._validate_methods = {}
+        owner._validate_methods[name] = {}
+
+        signature_info = inspect.getfullargspec(self.fn)
+
+        for curr_arg in self._args_to_validate:
+            owner._validate_methods[name][curr_arg] = signature_info.annotations[
+                curr_arg
+            ]
+
+        # then replace ourself with the original method
+        setattr(owner, name, curr_func)
+
+
+class InterfaceValidator:
+    def __init__(self, verbose: bool = False):
+        self.__dict__["_properly_initialized_InterfaceValidator"] = True
+        self._verbose_signature_validation = verbose
+        if not hasattr(self, "_validate_methods"):
+            raise Exception(
+                "InterfaceValidator: error - seems like you did not requested signature validation for any method. To use that you need to use @interface_validator.validate_signature on at least one method"
+            )
+
+        for method_name, curr_validate_args in self._validate_methods.items():
+            curr_bases = self.__class__.__bases__
+            if len(curr_bases) > 1:
+                raise Exception("currently not supporting more than one base")
+            if len(curr_bases) == 0:
+                raise Exception("expected at least one base")
+
+            if getattr(type(self), method_name) == getattr(curr_bases[0], method_name):
+                raise Exception(
+                    f'a class that inherited from InsilicoLabModel did not override interface method "{method_name}" !'
```

Comment:
Do you mean InterfaceValidator?

Reply:
yes, it's leftover from before I extracted this class, thanks! fixed